File: app/src/main/java/com/bzh/dytt/search/SearchFragment.java
package com.bzh.dytt.search;


import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.WindowManager;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.TextView;

import com.bzh.dytt.R;
import com.bzh.dytt.data.VideoDetail;
import com.bzh.dytt.data.network.Resource;
import com.bzh.dytt.home.MovieListAdapter;
import com.bzh.dytt.home.SingleListFragment;

import java.util.List;

import javax.inject.Inject;

public class SearchFragment extends SingleListFragment<VideoDetail> {

    private static final String TAG = "SearchFragment";

    @Inject
    ViewModelProvider.Factory mViewModelFactory;
    @Inject
    ViewModelProvider.Factory mFactory;
    private EditText mSearchInput;
    private TextView.OnEditorActionListener mSearchActionListener = new TextView.OnEditorActionListener() {
        @Override
        public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
            if (actionId == EditorInfo.IME_ACTION_SEARCH) {
                if (getActivity() != null && !TextUtils.isEmpty(v.getText())) {
                    String searchTarget = v.getText().toString().trim();
                    ((SearchViewModel) mViewModel).setQuery(searchTarget);
                    mSwipeRefresh.setEnabled(true);
                    mSwipeRefresh.setRefreshing(true);
                    InputMethodManager imm = (InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE);
                    if (imm != null) {
                        imm.hideSoftInputFromWindow(mSearchInput.getWindowToken(), InputMethodManager.HIDE_IMPLICIT_ONLY);
                    }
                }
                return true;
            }
            return false;
        }
    };

    public static SearchFragment newInstance() {

        Bundle args = new Bundle();

        SearchFragment fragment = new SearchFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    protected void doCreate(@Nullable Bundle savedInstanceState) {
        if (getActivity() != null && getActivity().getWindow() != null) {
            int softInputMode = WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE | WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN;
            getActivity().getWindow().setSoftInputMode(softInputMode);
        }
        super.doCreate(savedInstanceState);
        setupActionBar();
    }

    private void setupActionBar() {
        if (getActivity() != null && getActivity() instanceof AppCompatActivity) {
            ActionBar actionBar = ((AppCompatActivity) getActivity()).getSupportActionBar();
            if (actionBar != null) {
                actionBar.setDisplayShowTitleEnabled(false);
                actionBar.setDisplayShowCustomEnabled(true);
                ActionBar.LayoutParams lp1 = new ActionBar.LayoutParams(ActionBar.LayoutParams.MATCH_PARENT, ActionBar.LayoutParams.MATCH_PARENT);
                View customNav = LayoutInflater.from(getActivity()).inflate(R.layout.search_action_bar, null); // layout which contains your button.
                actionBar.setCustomView(customNav, lp1);
                mSearchInput = customNav.findViewById(R.id.search_edit_input);
            }
        }
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mSearchInput.setOnEditorActionListener(mSearchActionListener);
        mSwipeRefresh.setEnabled(false);
        mSwipeRefresh.setRefreshing(false);
    }

    @Override
    protected RecyclerView.Adapter createAdapter() {
        return new MovieListAdapter(this.getContext());
    }

    @Override
    protected void replace(List<VideoDetail> listData) {
        ((MovieListAdapter) mAdapter).replace(listData);
    }

    @Override
    protected LiveData<Resource<List<VideoDetail>>> getListLiveData() {
        return ((SearchViewModel) mViewModel).getVideoList();
    }

    @Override
    protected LiveData<Throwable> getThrowableLiveData() {
        return ((SearchViewModel) mViewModel).getFetchVideoDetailState();
    }

    @Override
    public Observer<Resource<List<VideoDetail>>> getListObserver() {
        return new Observer<Resource<List<VideoDetail>>>() {
            @Override
            public void onChanged(@Nullable Resource<List<VideoDetail>> result) {
                mListObserver.onChanged(result);
                assert result != null;
                switch (result.status) {
                    case ERROR:
                    case SUCCESS:
                        mSwipeRefresh.setEnabled(false);
                        mSwipeRefresh.setRefreshing(false);
                        break;
                }
            }
        };
    }

    @Override
    protected ViewModel createViewModel() {
        return ViewModelProviders.of(this, mFactory).get(SearchViewModel.class);
    }
}

File: app/src/main/java/com/bzh/dytt/ui/search/SearchFragment.kt
package com.bzh.dytt.ui.search

import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import android.content.Context
import androidx.databinding.DataBindingUtil
import android.os.Bundle
import androidx.fragment.app.FragmentActivity
import androidx.appcompat.app.ActionBar
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.AsyncDifferConfig
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.text.TextUtils
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.TextView
import com.bzh.dytt.AppExecutors
import com.bzh.dytt.R
import com.bzh.dytt.base.BaseFragment
import com.bzh.dytt.databinding.SearchListPageBinding
import com.bzh.dytt.ui.home.HomeListAdapter
import com.bzh.dytt.ui.home.MovieItemHolder
import com.bzh.dytt.util.autoCleared
import com.bzh.dytt.vo.MovieDetail
import kotlinx.android.synthetic.main.home_list_page.*
import javax.inject.Inject


class SearchFragment : BaseFragment() {

    private var binding by autoCleared<SearchListPageBinding>()

    @Inject
    lateinit var appExecutors: AppExecutors

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    lateinit var viewModel: SearchViewModel

    lateinit var linearLayoutManager: androidx.recyclerview.widget.LinearLayoutManager

    lateinit var adapter: SearchListAdapter

    lateinit var editText: EditText

    private val refreshObserver: Observer<Boolean> = Observer {
        if (it != null) {
            swipe_refresh_layout.isEnabled = it
            swipe_refresh_layout.isRefreshing = it
        } else {
            swipe_refresh_layout.isEnabled = false
            swipe_refresh_layout.isRefreshing = false
        }
    }

    private val searchActionListener: TextView.OnEditorActionListener = TextView.OnEditorActionListener { v, actionId, _ ->
        if (actionId == EditorInfo.IME_ACTION_SEARCH) {
            if (!TextUtils.isEmpty(v.text)) {
                val searchTarget = v.text.toString().trim()
                viewModel.setQuery(searchTarget)

                val imm: InputMethodManager = activity?.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                imm.hideSoftInputFromWindow(editText.windowToken, InputMethodManager.HIDE_IMPLICIT_ONLY)
                return@OnEditorActionListener true
            }
        }
        return@OnEditorActionListener false
    }

    private var listObserver: Observer<List<MovieDetail>> = Observer { result ->
        adapter.submitList(result)
    }

    override fun doCreate(savedInstanceState: Bundle?) {
        val softInputMode = WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE or WindowManager.LayoutParams.SOFT_INPUT_ADJUST_PAN;
        activity?.window?.setSoftInputMode(softInputMode)
        super.doCreate(savedInstanceState)
        setupActionBar()
    }

    private fun setupActionBar() {
        if (activity is AppCompatActivity) {
            val actionBar = (activity as AppCompatActivity).supportActionBar
            if (actionBar != null) {
                actionBar.setDisplayShowTitleEnabled(false)
                actionBar.setDisplayShowCustomEnabled(true)
                val lp = ActionBar.LayoutParams(ActionBar.LayoutParams.MATCH_PARENT, ActionBar.LayoutParams.MATCH_PARENT)
                val customNav = LayoutInflater.from(activity).inflate(R.layout.search_action_bar, null)
                actionBar.setCustomView(customNav, lp)
                editText = customNav.findViewById(R.id.search_edit_input)
                editText.requestFocus()
            }
        }
    }

    override fun doCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = DataBindingUtil.inflate(inflater, R.layout.search_list_page, container, false)

        viewModel = viewModelFactory.create(SearchViewModel::class.java)
        lifecycle.addObserver(viewModel)
        binding.viewModel = viewModel
        viewModel.movieListLiveData.observe(this, listObserver)
        viewModel.refreshLiveData.observe(this, refreshObserver)

        return binding.root
    }

    override fun doViewCreated(view: View, savedInstanceState: Bundle?) {
        super.doViewCreated(view, savedInstanceState)

        editText.setOnEditorActionListener(searchActionListener)

        linearLayoutManager = androidx.recyclerview.widget.LinearLayoutManager(activity)
        binding.recyclerView.layoutManager = linearLayoutManager

        adapter = SearchListAdapter(activity, viewModel, appExecutors)
        binding.recyclerView.adapter = adapter
    }

    companion object {

        private const val TAG = "HomeListFragment"

        fun newInstance(): SearchFragment {
            val bundle = Bundle()
            val fragment = SearchFragment()
            fragment.arguments = bundle
            return fragment
        }
    }
}

class SearchListAdapter constructor(val activity: androidx.fragment.app.FragmentActivity?, val viewModel: SearchViewModel, appExecutors: AppExecutors) : ListAdapter<MovieDetail, MovieItemHolder>(
        AsyncDifferConfig
                .Builder<MovieDetail>(object : DiffUtil.ItemCallback<MovieDetail>() {
                    override fun areItemsTheSame(oldItem: MovieDetail, newItem: MovieDetail): Boolean {
                        return oldItem.id == newItem.id
                                && oldItem.categoryId == newItem.categoryId
                    }

                    override fun areContentsTheSame(oldItem: MovieDetail, newItem: MovieDetail): Boolean {
                        return oldItem.name == newItem.name
                                && oldItem.homePicUrl == newItem.homePicUrl
                    }
                })
                .setBackgroundThreadExecutor(appExecutors.diskIO())
                .build()
) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieItemHolder {
        return MovieItemHolder(activity,
                DataBindingUtil.inflate(
                        LayoutInflater.from(parent.context),
                        R.layout.item_home_child,
                        parent,
                        false
                )
        )
    }

    override fun onBindViewHolder(holder: MovieItemHolder, position: Int) {
        getItem(position).let { movieDetail ->
            Log.d(HomeListAdapter.TAG, "onBindViewHolder() called with: movieDetail = [${movieDetail.id} ${movieDetail.name}]")
            with(holder) {
                itemView.tag = movieDetail
                reset()
                bind(movieDetail)
            }
        }
    }

    override fun onViewAttachedToWindow(holder: MovieItemHolder) {
        super.onViewAttachedToWindow(holder)
        if (holder.adapterPosition != androidx.recyclerview.widget.RecyclerView.NO_POSITION) {
            getItem(holder.adapterPosition).let {
                viewModel.doUpdateMovieDetail(it)
            }
        }
    }

    override fun onViewDetachedFromWindow(holder: MovieItemHolder) {
        super.onViewDetachedFromWindow(holder)
        if (holder.adapterPosition != androidx.recyclerview.widget.RecyclerView.NO_POSITION) {
            getItem(holder.adapterPosition).let {
                viewModel.doRemoveUpdateMovieDetail(it)
            }
        }
    }
}
